### DISCUSSION TITLE
How can I build a distribution when `.git` is missing?

### DISCUSSION THREAD

**Discussion by brianthelion** [pyscaffold/pyscaffold#485] [Q&A]
At the bottom of the [FAQ](https://pyscaffold.org/en/stable/faq.html) is the following:

> ### How can I build a distribution if I have only the source code without a proper git repo?
> If you see an error message like:
>
> `setuptools-scm was unable to detect version for 'your/project'.`
>
> This means that `setuptools-scm` could not find an intact git repository. If you still want to build a distribution from the source code there is a workaround. In `setup.cfg` in the section `[metadata]` define a version manually with e.g. `version = 1.0`. Now remove from `pyproject.toml` the requirement `use_scm_version={"version_scheme": "no-guess-dev"}` if you use isolated builds with `tox` and/or `"setuptools_scm[toml]>=5"` from `setup.cfg` if you use `python setup.py bdist_wheel` to build.

I'm not having any luck getting this to work. Can someone provide a `bash` script (or similar) that shows the exact changes that need to be made for it to succeed?

**Comment by FlorianWilhelm** [comment]
Hi, can you post the error you are seeing? Are you using `tox` or `python setup.py bdist_wheel` to build? Which version of PyScaffold was used to set up the project?

**Reply by brianthelion** [reply]
Sorry, this turned out to be an issue with a failed `putup -U`. The following `Dockerfile` is currently working on v4.0.2:

```Dockerfile
# Dockerfile
FROM python:3.7-slim
COPY pyproject.toml setup.py setup.cfg ./
COPY src ./src
RUN sed -i 's/version_scheme/# version_scheme/g' pyproject.toml
RUN python setup.py bdist_wheel
```

**Reply by FlorianWilhelm** [reply]
Cool, so it works now?

**Reply by brianthelion** [reply]
Only in `docker`. I can't get the same recipe to work in my development `venv`.

**Reply by brianthelion** [reply]
In my development `venv`, the only way to get this to work is to make the following string replacements:

```yaml
setup.py:
      'use_scm_version={"version_scheme": "no-guess-dev"},': ""

pyproject.toml:
      '[tool.setuptools_scm]': ""
      'version_scheme = "no-guess-dev"': ""
```

**Reply by FlorianWilhelm** [reply]
okay, by doing so you deactivate setuptools-scm which does the git magic to retrieve the version. What is your git version btw? Maybe it's too old.

**Reply by brianthelion** [reply]
```bash
$ git --version
git version 2.27.0
```

**Reply by FlorianWilhelm** [reply]
okay, that's pretty new.

**Reply by FlorianWilhelm** [reply]
Can you run `git describe --dirty --tags --long --match "*[0-9]*"` in your git repo?

**Reply by brianthelion** [reply]
```bash
$ git describe --dirty --tags --long --match "*[0-9]*"
fatal: No names found, cannot describe anything.
```
Just to be clear, the `pyscaffold` project in question is nested within a mono-repo. I ran the above at the root of the mono-repo.

**Reply by FlorianWilhelm** [reply]
Okay, this might be the problem. You have no tags in your repo. Could you run `git fetch origin master --tags` (or `main`) to retrieve the tags from the server and see if it changes anything?

**Reply by brianthelion** [reply]
```bash
$ git fetch origin master --tags
From https://github.com/f0cal/my-package
 * branch            master     -> FETCH_HEAD
```

**Reply by FlorianWilhelm** [reply]
Could you add some tag with `git tag v0.1 -m "v0.1"` and see if that helps? And also please show the error output.

**Reply by brianthelion** [reply]
Sorry, just backing up here for a second: My goal is to perform the minimum number of modifications to make `pyscaffold` work _without_ a `.git` directory present. I can't `git tag` if there's no `.git` at hand.

Maybe you've changed the problem statement and I've failed to follow?

**Reply by FlorianWilhelm** [reply]
Okay, right but then it works as designed, right? You said:

```
In my development venv, the only way to get this to work is to make the following string replacements:

setup.py:
      'use_scm_version={"version_scheme": "no-guess-dev"},': ""

pyproject.toml:
      '[tool.setuptools_scm]': ""
      'version_scheme = "no-guess-dev"': "
```
 and that's exactly the solution. Depending on how you build, pip and setuptools version it might be that only the changes in `pyproject.toml` are enough but it's better to do what you did. 
 
In your dockerfile though, it seems that the package is built and it will get the version `unknown` right? This is why you should define `version = 1.0` in `setup.cfg` too as described in the FAQ. 

Is your questioned answered or are there still any open points?

**Reply by brianthelion** [reply]
No, with the Dockerfile, the package gets `-0.0.0`. So an open question is, why does the Dockerfile version require fewer modifications?

**Reply by FlorianWilhelm** [reply]
If you only do the change as in the Dockerfile. What is the error output?

**Comment by abravalheri** [comment]
Hi @brianthelion, I did a quick test here and I believe you can also use the [`SETUPTOOLS_SCM_PRETEND_VERSION`](https://github.com/pypa/setuptools_scm#environment-variables) environment variable inside the container to bypass the integration with `git`, e.g. `SETUPTOOLS_SCM_PRETEND_VERSION=0.1.0 python setup.py bdist_wheel`.
Please notice it might require one of the latest versions of `setuptuools-scm` (e.g. 6.3.1).

You might also want to have a look on [this old issue](https://github.com/pypa/setuptools_scm/issues/77). Things might have changed since it was filed, but in general `setuptools-scm` is very consistent between versions.

**Comment by julie777** [comment]
Slightly different, but I would like putup to at least run without git installed. If some user finds the project and installs in from PyPI just to see what it does and they run `putup <testproject>` it should notify them, but not fail if git is not installed or not configured.

When your first experience with a tool is that it doesn't work, you might not try again.
I realize to do more than just look at the directory structure and config files that you need other tools installed, and our prerequisites in the doc should list all of it.

I am thinking of someone who is doing some python programming already and sees pyscaffold on PyPI and just pip installs it to give it a try. They don't read any documentation they just try it. This has the added issue that we don't provide a runnable pyscaffold, which is the first thing a user will try to run because that works for most tools.  

(I just had this issue when trying to quickly test coloring output on a variety of platforms.)

**Reply by FlorianWilhelm** [reply]
Currently we state if git is not installed:
```
"Make sure git is installed and working. "
"Use flag --very-verbose for more information."
```
and if not configured:
```
"Make sure git is configured. Run:\n"
'  git config --global user.email "you@example.com"\n'
'  git config --global user.name "Your Name"\n'
"to set your account's default identity."
```
What do you mean with
>  it should notify them, but not fail if git is not installed or not configured.
What do you suggest instead? Do not return an error code unequal 0?

**Reply by FlorianWilhelm** [reply]
Also, could we have this suggestion in a discussion or even issue? It is not directly related to the question asked here. The question is about building the project and not running `putup` when git is missing.

**Comment by julie777** [comment]
We do notify them. I wasn't implying that we don't. I was just being complete. 
That we abort after the notification is was I was pointing out.

This could be a case where I am trying handle all possibilities, which might not be what we want to do.

**Reply by FlorianWilhelm** [reply]
Ok, understood. Instead of aborting what would you do? Create the template without git? Now I get it. 

The problem is that if we do this, many features might not work or break at some point if people then use such a partial scaffold.

**Reply by julie777** [reply]
> Ok, understood. Instead of aborting what would you do? Create the template without git? Now I get it.
> 
> The problem is that if we do this, many features might not work or break at some point if people then use such a partial scaffold.

It is true there could be some issues, although I can't imagine what by not having git available. In any case, we would be showing a warning and they could choose to install and set up git and recreate the project, or just use it as is.

**Reply by abravalheri** [reply]
That will definitely break the first time they try to build the package and I suspect also when they do a editable install (unless they workaround `setuptools-scm`).
